_practicalities/intro.md

```diff
@@ -0,0 +1,355 @@
+data pip and setuptools:
+
+
+```
+import sys
+
+if sys.version_info < (3,):
+    raise ImportError(
+    """You are running Frobulator 6.0 on Python 2
+
+Unfortunately Frobulator 6.0 and above are not compatible with Python 2
+anymore, and you still ended up with this version installed on your system.
+That's a bummer. Sorry about that. It should not have happen. Make sure you
+have pip >= 9.0 to avoid this kind of issues, as well as setuptools >= 24.2:
+
+ $ pip install pip setuptools --upgrade
+
+You have various other choices
+
+- install an older version of Frobulator:
+
+ $ pip install 'frobulator<6.0'
+
+- Upgrade your system to use Python 3.
+
+It would be great if you can figure out how this version ended up being
+installed, and try to check how to prevent that for future users.
+
+See the following url for more up to date informations:
+
+https://i.am.an/url
+
+""")
+
+```
+
+## Watch out for beta releases
+
+
+Make sure your version number match pep 440 or you will get surprises during
+beta in particular as the `sdist` and `wheel` will appear as being different
+versions, in particular sdist (during beta/rc/post) can appear with a greater
+version number than wheels. Pip thus try to install the sdist instead of the
+wheel, which have more chance of failing, in particular with pre 24.2 versions
+of setuptools.
+
+The regular expression to check for validity of pep440 can be find below:
+
+  `^([1-9]\\d*!)?(0|[1-9]\\d*)(\\.(0|[1-9]\\d*))*((a|b|rc)(0|[1-9]\\d*))?(\\.post(0|[1-9]\\d*))?(\\.dev(0|[1-9]\\d*))?`
+
+## Depend on setuptools
+
+You can mark your library as dependent on setuptools greater than 24.3 starting
+now, this will insure that during the next upgrade (when the packages drop
+python 2 support) will have the right version of setuptools.
+
+Of course regardless of all the care you will take for your library to no break
+and to install only on python 2, you will likely have cases where it still end
+up being installed on incompatible versions of Python. Simply because users
+upgrades rarely and only an old version of pip or setuptools is enough to make
+the all update process broken.
+
+## fail early in setup.py
+
+Leave `setup.py` python 2 compatible and fail early. If you detect Python 2
+raise a clear error message and ask user to make sure they have pip >9.0 (or
+migrate to Python 3). You can (try to) conditionally import pip and check for
+its version but this might not be the same pip. Failing early is important to
+make sure the Python installation does not install and incompatible version.
+Otherwise user code can fail at runtime arbitrary later in the future, which can
+be a difficult to debug and fix. Get inspiration from the message of failure at
+runtime, and adapt for installation time.
+
+## Fix dependant libraries
+
+If you control dependant packages, Make sure to include conditional dependencies
+depending on the version of Python.
+
+# Incorrect mitigations
```

Comment:
"Incorrect" seems a bit strong considering that the first one can work, and actually works on older versions of pip as well. Maybe "Non-recommended"?